accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,165 @@
+
+**Open issue:** should we do anything to make it easier to identify assemblies which might need an update?  We can consider some case studies.  
+
+If a specific compiler version has a code-gen bug it isn't clear from the final assembly if that bug is present, unless you directly inspect the IL/bytecode for the issue.  The compiler version is embedded in the PDB of the assembly which could be used to identify assemblies with the issue.  There is no notification at design time or runtime of any issue with the compiler, the user needs to subscribe to updates to understand when an update is needed.
+
+If a runtime assembly has a bug we have multiple avenues of detection.  We could examine the assembly metadata files contained in the application layout, including for self-contained applications.  The runtime will also write "breadcrumb" files on a machine when code is executed that is annotated for servicing that help identify that a specific assembly or package version has been executed on that machine.  Today we use these breadcrumbs to help target updates to a machine.  Those updates can install in a well known location and be preferred over the application's copy.
+
+So for means of detection we have:
+1. Manual
+2. PDB indicator: eg source generator assembly version, file version, hash)
+3. Assembly metadata in consuming user project: eg AssemblyMetadata("SourceGeneratedBy", "source generator assembly version, file version, hash")
+4. Runtime breadcrumbs
+
+For means of update we have:
+1. No notification, user needs to subscribe to servicing updates.
+2. Design time notification.
+3. Killbit in the runtime to prevent execution of user assemblies that contain vulnerable source generated code.
+4. ~~Automatically update user assembly though a hotpatch applied by runtime or JIT~~
+
+We'll need to better understand what customer's expectations are around servicing to decide if we should do any work on these axes.  My initial inclination is to only have a PDB indicator and require users to subscribe to updates.
+
+### Functionality concerns for source generators
+
+Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.
+
+Source generator assemblies should target .NETStandard and must run on both .NETFramework and .NETCore. Source generators should support generating source for all TargetFrameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the TargetFramework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.
```

Comment:
Is the .NET Standard targeting requirement coming from the fact that the compiler itself is .NET Standard 2.0-based?  So even if it's a generator for System.SomeCoolThingForNet6WithNoOlderOffering.dll-based components it will be invoked by the compiler, so is bound by the compiler's requirements?

(If so, it might seem more like "this is a technical constraint to be aware of" than "this is a decision of aesthetics/politics/something that you may want to argue with" to mention that)

Reply:
Correct, technical constraint.

Reply:
Clarifying this.